src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs

```diff
@@ -507,14 +507,32 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                 }
             }
 
-            private class OnnxRuntimeOutputCacher
+            private class OnnxRuntimeOutputCacher : IDisposable
             {
                 public long Position;
-                public Dictionary<string, NamedOnnxValue> Outputs;
+                public Dictionary<string, DisposableNamedOnnxValue> Outputs;
                 public OnnxRuntimeOutputCacher()
                 {
                     Position = -1;
-                    Outputs = new Dictionary<string, NamedOnnxValue>();
+                    Outputs = new Dictionary<string, DisposableNamedOnnxValue>();
+                }
+
+                private bool _isDisposed;
+                public void Dispose()
+                {
+                    if (_isDisposed)
+                        return;
+                    foreach (var onnxValue in Outputs.Values)
+                        onnxValue.Dispose();
+                    _isDisposed = true;
+                }
+
+                ~OnnxRuntimeOutputCacher()
+                {
+                    if (_isDisposed)
```

Comment:
Does it ever hit this code path?

Reply:
The finalizer is hit. I dont think the Dispose call is though. I left it in cause it would be better to switch to a dispose call when we are able to.

Reply:
We shouldn't be using a finalizer here.

1. We should ensure `OnnxRuntimeOutputCacher` instances get disposed properly.
2. In the cases where `Dispose()` is never called by the user, and this object is no longer referenced, all the references to the underlying `DisposableNamedOnnxValue` will be removed as well. If `DisposableNamedOnnxValue` is holding on to native memory, it should have a finalizer (or more preferrably use a SafeHandle). The managed object that manages the native memory is responsible for ensuring proper clean up.

We shouldn't be using finalizers to dispose managed objects. Finalizers should only be used for managing native resources.